### PR TITLE
Switch to NGINX by default, make Apache and Lighttpd optional

### PR DESCRIPTION
**Pull request by andi34**
Thanks to @rawbertp .

By default NGINX is used.
To use Apache run `sudo bash install-raspbian.sh apache` ,
To use Lighttpd run `sudo bash install-raspbian.sh lighttpd`.

If PHP can not be enabled in config for NGINX or Lighttpd we'll use Apache as fallback as we don't need to enable PHP via config file.
Fix https://github.com/andreknieriem/photobooth/issues/177

### CODE REVIEW

File: install-raspbian.sh [hunk 1]
```diff
@@ -67,7 +67,8 @@ apt update
 apt dist-upgrade -y
 
 info "### Photobooth needs some software to run."
-apt install -y libapache2-mod-php php-gd gphoto2 unclutter
+apt install -y nginx php-fpm php-gd gphoto2 unclutter
+/etc/init.d/nginx start
```
Comment (rawbertp) [inline install-raspbian.sh]: `systemctl start nginx` should be used but it's not necessary here as the service will be automatically started.

File: install-raspbian.sh [hunk 2]
```diff
@@ -114,6 +115,29 @@ else
...
+    info "### Enable PHP in NGINX"
+    cp /etc/nginx/sites-enabled/default ~/nginx-default.bak
+    sed -i 's/^\(\s*\)index index\.html\(.*\)/\1index index\.php index\.html\2/g' /etc/nginx/sites-available/default
+    sed -i '/location ~ \\.php$ {/s/^\(\s*\)#/\1/g' /etc/nginx/sites-available/default
+    sed -i '/include snippets\/fastcgi-php.conf/s/^\(\s*\)#/\1/g' /etc/nginx/sites-available/default
+    sed -i '/fastcgi_pass unix:\/run\/php\//s/^\(\s*\)#/\1/g' /etc/nginx/sites-available/default
+    sed -i '/.*fastcgi_pass unix:\/run\/php\//,// { /}/s/^\(\s*\)#/\1/g; }' /etc/nginx/sites-available/default
+
+    info "### Testing NGINX config"
+    /usr/sbin/nginx -t -c /etc/nginx/nginx.conf &>/dev/null && RESULT='config test ok' || RESULT='config test failed'
```
Comment (rawbertp) [inline install-raspbian.sh]: This should be changed to:
```bash
/usr/sbin/nginx -t -c /etc/nginx/nginx.conf
systemctl reload nginx
```
... as the test run will output useful information what is wrong with the config.

File: install-raspbian.sh [hunk 3]
```diff
@@ -114,6 +115,29 @@ else
...
+
+    if [[ $RESULT == 'config test ok' ]]; then
+        info "$RESULT"
+        info "### Restarting NGINX"
+        /etc/init.d/nginx restart
+    else
+        error "$RESULT"
+    fi
```
Comment (rawbertp) [inline install-raspbian.sh]: No need for this check. If the rest run fails the script will abort anyway, due to `set -e`.

File: install-raspbian.sh [hunk 4]
```diff
@@ -114,6 +114,24 @@ else
     tar -xzvf /tmp/photobooth-latest.tar.gz -C /var/www/html/
 fi
 
+if [ -f "/etc/nginx/sites-enabled/default" ]; then
```
Comment (rawbertp) [inline install-raspbian.sh]: I think it should be okay from an Nginx perspective now. 👍 

One thing I personally would change, is to put the path to the nginx-config into a variable:

```bash
nginx_conf="/etc/nginx/sites-enabled/default"
if [ -f "${nginx_conf}" ]; then
  cp "${nginx_conf}" ...
  sed -i '...' "${nginx_conf}"
...
```
	

```

Comment (andi34) [reply install-raspbian.sh]: Thanks, just shutdown my PC. I'll update tomorrow evening :)

Comment (rawbertp) [reply install-raspbian.sh]: That's no big deal, I just don't like "code duplication" a lot (I know in my example I did it like that :p).

**Comment by sualko** [comment]
A performance comparison would be nice.

**Comment by andi34** [comment]
> A performance comparison would be nice.

By reverting https://github.com/andreknieriem/photobooth/commit/b7113f82d201aa413f22ce09dd3c707a37629750 or anything other special?

**Comment by andi34** [comment]
> A performance comparison would be nice.

By reverting https://github.com/andreknieriem/photobooth/commit/b7113f82d201aa413f22ce09dd3c707a37629750 or anything other special?

**Comment by andi34** [comment]
> A performance comparison would be nice.

By reverting https://github.com/andreknieriem/photobooth/commit/b7113f82d201aa413f22ce09dd3c707a37629750 or anything other special?

File: install-raspbian.sh (hunk 4 above)
Comment (andi34) [reply install-raspbian.sh]:
```
diff --git a/install-raspbian.sh b/install-raspbian.sh
index ac4a69c..4086667 100755
--- a/install-raspbian.sh
+++ b/install-raspbian.sh
@@ -114,14 +114,16 @@ else
     tar -xzvf /tmp/photobooth-latest.tar.gz -C /var/www/html/
 fi
 
-if [ -f "/etc/nginx/sites-enabled/default" ]; then
+nginx_conf="/etc/nginx/sites-enabled/default"
+
+if [ -f "${nginx_conf}" ]; then
     info "### Enable PHP in NGINX"
-    cp /etc/nginx/sites-enabled/default ~/nginx-default.bak
-    sed -i 's/^\(\s*\)index index\.html\(.*\)/\1index index\.php index\.html\2/g' /etc/nginx/sites-available/default
-    sed -i '/location ~ \\.php$ {/s/^\(\s*\)#/\1/g' /etc/nginx/sites-available/default
-    sed -i '/include snippets\/fastcgi-php.conf/s/^\(\s*\)#/\1/g' /etc/nginx/sites-available/default
-    sed -i '/fastcgi_pass unix:\/run\/php\//s/^\(\s*\)#/\1/g' /etc/nginx/sites-available/default
-    sed -i '/.*fastcgi_pass unix:\/run\/php\//,// { /}/s/^\(\s*\)#/\1/g; }' /etc/nginx/sites-available/default
+    cp "${nginx_conf}" ~/nginx-default.bak
+    sed -i 's/^\(\s*\)index index\.html\(.*\)/\1index index\.php index\.html\2/g' "${nginx_conf}"
+    sed -i '/location ~ \\.php$ {/s/^\(\s*\)#/\1/g' "${nginx_conf}"
+    sed -i '/include snippets\/fastcgi-php.conf/s/^\(\s*\)#/\1/g' "${nginx_conf}"
+    sed -i '/fastcgi_pass unix:\/run\/php\//s/^\(\s*\)#/\1/g' "${nginx_conf}"
+    sed -i '/.*fastcgi_pass unix:\/run\/php\//,// { /}/s/^\(\s*\)#/\1/g; }' "${nginx_conf}"
 
     info "### Testing NGINX config"
     /usr/sbin/nginx -t -c /etc/nginx/nginx.conf
@@ -129,7 +131,7 @@ if [ -f "/etc/nginx/sites-enabled/default" ]; then
     info "### Restarting NGINX"
     systemctl reload nginx
 else
-    error "Can not find /etc/nginx/sites-enabled/default"
+    error "Can not find ${nginx_conf}!"
 fi
 
 info "### Setting permissions."
```

**Comment by sualko** [comment]
> By reverting b7113f8 or anything other special?

I think that wood be nice. Maybe also a comparison to Lighttpd.

**Comment by andi34** [comment]
Compared NGINX and Apache, posted here https://github.com/andreknieriem/photobooth/issues/177

Out of time today to test lighttpd too.

**Comment by andi34** [comment]
Install Apache instead NGINX by running `sudo bash install-raspbian.sh apache`

**Comment by andi34** [comment]
We could also handle to use `lighttpd` this way. At least would give all options to the end-user.

I'll check lighttpd too next days, maybe I've a little time to at least test it's working too.

**Comment by andi34** [comment]
To use lighttpd
```
sudo apt install -y lighttpd php-fpm php-gd git gphoto2
sudo lighttpd-enable-mod fastcgi
sudo lighttpd-enable-mod fastcgi-php
sudo nano /etc/lighttpd/conf-available/15-fastcgi-php.conf 
sudo cp /etc/lighttpd/conf-available/15-fastcgi-php.conf /etc/lighttpd/conf-available/15-fastcgi-php.conf.bak
sudo nano /etc/lighttpd/conf-available/15-fastcgi-php.conf 
```
change from
```
# -*- depends: fastcgi -*-
# /usr/share/doc/lighttpd/fastcgi.txt.gz
# http://redmine.lighttpd.net/projects/lighttpd/wiki/Docs:ConfigurationOptions#mod_fastcgi-fastcgi

## Start an FastCGI server for php (needs the php5-cgi package)
fastcgi.server += ( ".php" => 
	((
		"bin-path" => "/usr/bin/php-cgi",
		"socket" => "/var/run/lighttpd/php.socket",
		"max-procs" => 1,
		"bin-environment" => ( 
			"PHP_FCGI_CHILDREN" => "4",
			"PHP_FCGI_MAX_REQUESTS" => "10000"
		),
		"bin-copy-environment" => (
			"PATH", "SHELL", "USER"
		),
		"broken-scriptfilename" => "enable"
	))
)
```
to 
```
# -*- depends: fastcgi -*-
# /usr/share/doc/lighttpd/fastcgi.txt.gz
# http://redmine.lighttpd.net/projects/lighttpd/wiki/Docs:ConfigurationOptions#mod_fastcgi-fastcgi

## Start an FastCGI server for php (needs the php5-cgi package)
fastcgi.server += ( ".php" => 
	((
		"socket" => "/var/run/php/php7.3-fpm.sock",
		"broken-scriptfilename" => "enable"
	))
)
```

Reload the service:
```
sudo service lighttpd force-reload
```

File: install-raspbian.sh [hunk 5]
```diff
@@ -31,6 +31,69 @@ if [[ $PI_MODEL != Raspberry* ]]; then
...
+        cat > ${php_conf} <<EOF
+fastcgi.server += ( ".php" => 
+	((
+		"socket" => "/var/run/php/php7.3-fpm.sock",
+		"broken-scriptfilename" => "enable"
+	))
+)
+EOF
+
+    service lighttpd force-reload
```
Comment (andi34) [inline install-raspbian.sh]: More spaces

File: install-raspbian.sh [hunk 6]
```diff
@@ -31,6 +31,74 @@ if [[ $PI_MODEL != Raspberry* ]]; then
...
+        error "Can not find ${nginx_conf} !"
+        info "Using Apache Webserver !"
+        apt remove nginx php-fpm
+        apache_webserver
+
+    fi
+}
+
+lighttpd_webserver() {
+    apt install -y lighttpd php-fpm
```
Comment (andi34) [inline install-raspbian.sh]: Add info `"### Installing Lighttpd Webserver..."`

File: install-raspbian.sh [hunk 7]
```diff
@@ -31,6 +31,75 @@ if [[ $PI_MODEL != Raspberry* ]]; then
...
+
+        info "### Testing NGINX config"
+        /usr/sbin/nginx -t -c /etc/nginx/nginx.conf
+
+        info "### Restarting NGINX"
+        systemctl reload nginx
+    else
+        error "Can not find ${nginx_conf} !"
+        info "Using Apache Webserver !"
+        apt remove nginx php-fpm
```
Comment (andi34) [inline install-raspbian.sh]: apt remove -y

File: install-raspbian.sh [hunk 8]
```diff
@@ -31,6 +31,75 @@ if [[ $PI_MODEL != Raspberry* ]]; then
...
+		"broken-scriptfilename" => "enable"
+	))
+)
+EOF
+
+        service lighttpd force-reload
+    else
+        error "Can not find ${php_conf} !"
+        info "Using Apache Webserver !"
+        apt remove lighttpd php-fpm
```
Comment (andi34) [inline install-raspbian.sh]: apt remove -y

**Comment by andi34** [comment]
All 3 Webserver installations tested with this script, all working fine.

**Comment by sualko** [comment]
I had no time to test it, but looks good. Even if I'm not sure if it's helpful to give the user so many possibilities. The script was intended for inexperienced users and it's quite hard to make a decision here.

**Pull request merged**